bolt-sidecar/src/onchain_registry.rs

```diff
@@ -0,0 +1,163 @@
+        let hashes = BoltValidators::pubkey_hashes(keys);
+
+        let returndata = self.0.getProposerStatuses(hashes).call().await;
+
+        // TODO: clean this after https://github.com/alloy-rs/alloy/issues/787 is merged
+        let error = match returndata.map(|data| data.statuses) {
+            Ok(statuses) => {
+                for status in &statuses {
+                    if !status.active {
+                        bail!(
+                            "validator with public key hash {:?} is not active in Bolt",
+                            status.pubkeyHash
+                        );
+                    }
+                }
+
+                return Ok(statuses);
+            }
+            Err(error) => match error {
+                ContractError::TransportError(TransportError::ErrorResp(err)) => {
+                    let data = err.data.unwrap_or_default();
+                    let data = data.get().trim_matches('"');
+                    let data = Bytes::from_str(data).unwrap_or_default();
+
+                    BoltManagerContractErrors::abi_decode(&data, true)?
+                }
+                e => return Err(e)?,
+            },
+        };
+
+        if matches!(error, BoltManagerContractErrors::ValidatorDoesNotExist(_)) {
+            // TODO: improve this error message once https://github.com/chainbound/bolt/issues/338
+            // is solved
+            bail!("not all validators are registered in Bolt");
+        } else {
+            Err(SolError::custom(format!(
+                "unexpected Solidity error selector: {:?}",
+                error.selector()
+            ))
+            .into())
+        }
+    }
+}
+
+/// Utility functions related to the BoltValidators contract
+pub struct BoltValidators;
+
+impl BoltValidators {
+    /// Hash the public keys of the proposers. This follows the same
+    /// implementation done on-chain in the BoltValidators contract.
+    pub fn pubkey_hashes(keys: &[BlsPublicKey]) -> Vec<B256> {
+        keys.iter().map(Self::pubkey_hash).collect()
+    }
+
+    /// Hash the public key of the proposer. This follows the same
+    /// implementation done on-chain in the BoltValidators contract.
+    pub fn pubkey_hash(key: &BlsPublicKey) -> B256 {
+        let digest = Self::pubkey_hash_digest(key);
+        keccak256(digest)
+    }
+
+    fn pubkey_hash_digest(key: &BlsPublicKey) -> B512 {
+        let mut onchain_pubkey_repr = B512::ZERO;
+
+        // copy the pubkey bytes into the rightmost 48 bytes of the 512-bit buffer.
+        // the result should look like this:
+        //
+        // 0x00000000000000000000000000000000b427fd179b35ef085409e4a98fb3ab84ee29c689df5c64020eab0b20a4f91170f610177db172dc091682df627c9f4021
+        // |<---------- 16 bytes ---------->||<----------------------------------------- 48 bytes ----------------------------------------->|
+        onchain_pubkey_repr[16..].copy_from_slice(key);
+        onchain_pubkey_repr
+    }
+}
+
+sol! {
```

Comment:
In the future we should probably do something to keep these interfaces up to date, @coderabbitai please draft an issue

Reply:
Beast